uri/uri.h

```diff
@@ -36,6 +40,87 @@ struct Uri {
     [[nodiscard]] bool operator==(Uri const &) const = default;
 };
 
+enum HostType { DNS_DOMAIN, IP4_ADDR, IP6_ADDR, OPAQUE, EMPTY };
+
+struct Host {
+    HostType type;
+
+    // union {
+    std::string domain;
+    std::uint32_t ip4_addr;
+    std::uint16_t ip6_addr[8];
+    //};
+};
+
+struct Origin {
+    std::string scheme;
+    Host *host;
+    std::optional<uint16_t> port;
+    std::optional<std::string> domain;
+    // Need this placeholder until I figure out what "opaqueness" means for an origin in this context
+    bool opaque;
+};
+
+enum ParserState {
+    SCHEME_START,
+    SCHEME,
+    NO_SCHEME,
+    SPECIAL_RELATIVE_OR_AUTHORITY,
+    PATH_OR_AUTHORITY,
+    RELATIVE,
+    RELATIVE_SLASH,
+    SPECIAL_AUTHORITY_SLASHES,
+    SPECIAL_AUTHORITY_IGNORE_SLASHES,
+    AUTHORITY,
+    HOST,
+    HOSTNAME,
+    PORT,
+    FILE,
+    FILE_SLASH,
+    FILE_HOST,
+    PATH_START,
+    PATH,
+    OPAQUE_PATH,
+    QUERY,
+    FRAGMENT,
+    FAILURE,
+    TERMINATE
+};
+
+inline const std::map<std::string, std::string> special_schemes = {
+        {"ftp", "21"}, {"file", ""}, {"http", "80"}, {"https", "443"}, {"ws", "80"}, {"wss", "443"}};
+
+/**
+ * Generates a new Blob URL for the given origin
+ */
+std::string blob_url_create(Origin *origin);
+
+class URL final : util::BaseParser {
```

Comment:
`Url` instead of `URL` to follow naming conventions.

I think we should keep the Url and the UrlParser as separate classes. Having a Url be a Parser feels like very weird design?

Reply:
idk, it seems ok to me, but I guess it would be fine to split them out.